eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -278,4 +346,116 @@ where
+        let c = c1.c + c2.c;
+        Ciphertext { c }
+    }
+
+    pub fn ciphertext_mul(&self, c1: Ciphertext<C>, c2: Ciphertext<C>)-> Ciphertext<C> {
+        let mut res = Poly::<C>::zero();
+        // lift to allow bignum coefficients (n * q * q would be enough, as in the C++ implementation)
+        let coeffs1 = c1.c.to_field_elements().unwrap();
+        let coeffs2 = c2.c.to_field_elements().unwrap();
+
+        let mut lifted_pol1 = DensePolynomial::<Fq79bn>::zero();
+        for i in 0..coeffs1.len() {
+            if i >= lifted_pol1.coeffs.len() {
+                lifted_pol1.coeffs.resize(i+1, C::coeff_as_u128(coeffs1[i]).into());
+            } else {
+                lifted_pol1[i] = C::coeff_as_u128(coeffs1[i]).into();
+            }
+        }
+        let mut lifted_pol2 = DensePolynomial::<Fq79bn>::zero();
+        for i in 0..coeffs2.len() {
+            if i >= lifted_pol2.coeffs.len() {
+                lifted_pol2.coeffs.resize(i+1, C::coeff_as_u128(coeffs2[i]).into());
+            } else {
+                lifted_pol2[i] = C::coeff_as_u128(coeffs2[i]).into();
+            }
+        }
+
+        // TODO: use more efficient version
+        let mut c = lifted_pol1.naive_mul(&lifted_pol2);
+        
+        // reduce by the cyclotomic polynomial
+        let mut i = C::MAX_POLY_DEGREE;
+        while i < c.coeffs.len() {
+            let q = i / C::MAX_POLY_DEGREE;
+            let r = i % C::MAX_POLY_DEGREE;
+
+            if q % 2 == 1 {
+                c.coeffs[r] = c.coeffs[r] - c.coeffs[i];
+            } else {
+                c.coeffs[r] = c.coeffs[r] + c.coeffs[i];
+            }
+            c.coeffs[i] = Fq79bn::zero();
+            i += 1;
+        }
+
+        // mul by t/q 
+        // round to the nearest int
+        // reduce mod q
+        // convert back to Fq79
+
+        for i in 0..c.coeffs.len() {
+
+            let mut cbn: num_bigint::BigInt = num_bigint::BigInt::from_bytes_le(num_bigint::Sign::Plus, &c[i].into_bigint().to_bytes_le());
+            let cbnd2: num_bigint::BigInt = num_bigint::BigInt::from_bytes_le(num_bigint::Sign::Plus, &Fq79bn::MODULUS_MINUS_ONE_DIV_TWO.to_bytes_le());
+            if cbn > cbnd2 {
+                let bnmod: num_bigint::BigInt = num_bigint::BigInt::from_bytes_le(num_bigint::Sign::Plus, &Fq79bn::MODULUS.to_bytes_le());
+                cbn -= bnmod;
+            }
+            cbn *= num_bigint::BigInt::from(C::t_as_u128());
+            let modbn = num_bigint::BigInt::from(C::modulus_as_u128());
+            let modd2bn = num_bigint::BigInt::from(C::modulus_minus_one_div_two_as_u128());
+            if cbn > num_bigint::BigInt::zero() {
+                cbn += modd2bn;
+            } else {
+                cbn -= modd2bn;
+            }
+            cbn /= modbn.clone(); 
+            cbn %= modbn.clone();
+            if cbn < num_bigint::BigInt::zero() {
+                cbn += modbn;
+            }
```

Comment:
Remove sign (this is probably redundant, conversion to `Fq` should do it anyway)